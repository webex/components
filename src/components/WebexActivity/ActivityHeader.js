import React from 'react';
import PropTypes from 'prop-types';
import {format, isToday, isSameWeek, isYesterday} from 'date-fns';

import WebexAvatar from '../WebexAvatar/WebexAvatar';
import {usePerson} from '../hooks';

import './ActivityHeader.scss';

/**
 * Returns a formatted timestamp based on the given date's offset from the current time.
 *
 * @param {Date} timestamp Date instance to format
 * @returns {string} formattedDate
 */
export function formatMessageDate(timestamp) {
  let formattedDate;

  if (isToday(timestamp)) {
    // 12:00 PM
    formattedDate = format(timestamp, 'p');
  } else if (isYesterday(timestamp)) {
    // Yesterday 12:00 PM
    formattedDate = `Yesterday ${format(timestamp, 'p')}`;
  } else if (isSameWeek(timestamp, new Date())) {
    // Monday 12:00 PM
    formattedDate = format(timestamp, 'iiii p');
  } else {
    // 1/1/2020 12:00 PM
    formattedDate = format(timestamp, 'M/d/yy p');
  }

  return formattedDate;
}

export default function ActivityHeader({personID, created}) {
  const {displayName} = usePerson(personID);

  return (
    <div className="activity-header">
      <WebexAvatar personID={personID} />
      <div className="activity-author">
        <span>{displayName}</span>
        <span className="activity-timestamp">{formatMessageDate(new Date(created))}</span>
      </div>
    </div>
  );
}

ActivityHeader.propTypes = {
  personID: PropTypes.string.isRequired,
  created: PropTypes.string.isRequired,
};
