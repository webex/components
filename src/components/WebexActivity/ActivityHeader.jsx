import React from 'react';
import PropTypes from 'prop-types';
import {
  format,
  isToday,
  isSameWeek,
  isYesterday,
} from 'date-fns';

import WebexAvatar from '../WebexAvatar/WebexAvatar';
import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {usePerson} from '../hooks';

/**
 * Returns a formatted timestamp based on the given date's offset from the current time.
 *
 * @param {Date} timestamp Date instance to format
 * @returns {string} formattedDate
 */
export function formatMessageDate(timestamp) {
  let formattedDate;

  if (isToday(timestamp)) {
    // 12:00 PM
    formattedDate = format(timestamp, 'p');
  } else if (isYesterday(timestamp)) {
    // Yesterday, 12:00 PM
    formattedDate = `Yesterday, ${format(timestamp, 'p')}`;
  } else if (isSameWeek(timestamp, new Date())) {
    // Monday, 12:00 PM
    formattedDate = format(timestamp, 'iiii, p');
  } else {
    // 1/1/2020, 12:00 PM
    formattedDate = format(timestamp, 'P, p');
  }

  return formattedDate;
}

/**
 * ActivityHeader component displays the header content of an activity.
 * Header includes the avatar of the activity author as well as the
 * activity timestamp.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function ActivityHeader({personID, timestamp}) {
  const {displayName} = usePerson(personID);

  return (
    <div className={`${WEBEX_COMPONENTS_CLASS_PREFIX}-activity-header`}>
      <WebexAvatar personID={personID} />
      <div className="activity-author">
        <span>{displayName}</span>
        <span className="activity-timestamp">{formatMessageDate(new Date(timestamp))}</span>
      </div>
    </div>
  );
}

ActivityHeader.propTypes = {
  personID: PropTypes.string.isRequired,
  timestamp: PropTypes.string.isRequired,
};
