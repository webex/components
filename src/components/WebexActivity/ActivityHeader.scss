$content-left-margin: 3.5rem; //avatar width + margin

.activity-header {
  display: flex;
  flex-direction: column;
  position: relative;

  .activity-author {
    @include noselect();
    display: flex;
    min-width: 18.75rem !important;
    margin-left: $content-left-margin;
    font-size: 0.875rem;
    line-height: 1.375rem;
    text-align: left;
    color: $md-gray-70;
    word-wrap: break-word;
  }

  .md-avatar {
    position: absolute;
    margin: 0 0.5rem;
  }

  .activity-timestamp {
    margin-left: 0.5rem;
  }
}
