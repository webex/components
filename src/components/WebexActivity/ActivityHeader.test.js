import {
  addDays,
  addHours,
  startOfToday,
  startOfWeek,
  startOfYesterday,
  subDays,
} from 'date-fns';

import {formatMessageDate} from './ActivityHeader';

describe('Activity Header', () => {
  let today;
  let sunday;

  beforeEach(() => {
    today = startOfToday();
    sunday = startOfWeek(today);
  });

  afterEach(() => {
    today = null;
    sunday = null;
  });

  describe('formatMessageDate()', () => {
    test('returns AM/PM time from today date', () => {
      const nineAM = addHours(today, 9);

      expect(formatMessageDate(nineAM)).toEqual('9:00 AM');
    });

    test('returns "Yesterday" and AM/PM time from yesterday date', () => {
      const yesterdayOnePM = addHours(startOfYesterday(), 13);

      expect(formatMessageDate(yesterdayOnePM)).toEqual('Yesterday, 1:00 PM');
    });

    test('returns day name and AM/PM time from a date of this week', () => {
      const monday = addDays(sunday, 1);
      const mondayTenAM = addHours(monday, 10);

      expect(formatMessageDate(mondayTenAM)).toEqual('Monday, 10:00 AM');
    });

    test('returns formatted date for dates older than a week', () => {
      const lastFriday = subDays(sunday, 2);
      const lastFridayNoon = addHours(lastFriday, 12);

      expect(formatMessageDate(lastFridayNoon)).toEqual('07/24/2020, 12:00 PM');
    });
  });
});
