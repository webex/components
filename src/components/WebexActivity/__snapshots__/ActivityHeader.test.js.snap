// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity Header component snapshot matches header at 3:00 PM 1`] = `
<div
  className="activity-header"
>
  <WebexAvatar
    displayStatus={true}
    personID="default"
  />
  <div
    className="activity-author"
  >
    <span>
      Webex Component User
    </span>
    <span
      className="activity-timestamp"
    >
      3:00 PM
    </span>
  </div>
</div>
`;
